java/src/jmri/jmrix/openlcb/OlcbProgrammer.java

```diff
@@ -46,62 +57,302 @@
+        checkProgramTrack();
+        getInterface().getMemoryConfigurationService().requestRead(nid, SPACE_DCC_CV, getCvAddress(CV), 1, new MemoryConfigurationService.McsReadHandler() {
+            @Override
+            public void handleReadData(NodeID nodeID, int i, long l, byte[] bytes) {
+                if (bytes.length < 1) {
+                    handleFailure(0x1000);
+                    return;
+                }
+                if (p != null) {
+                    notifyProgListenerEnd(p, bytes[0] & 0xff, ProgListener.OK);
+                }
+            }
+
+            @Override
+            public void handleFailure(int i) {
+                log.debug("CV {} read - memory config error 0x{}", CV, Integer.toHexString(i));
+                notifyProgListenerEnd(p, 0, olcbErrorToProgStatus(i));
+            }
+        });
     }
 
-    /** 
+    /**
      * {@inheritDoc}
      */
     @Override
-    public void confirmCV(String CV, int val, ProgListener p) {
+    public void confirmCV(String CV, int val, ProgListener p) throws ProgrammerException {
+        checkProgramTrack();
     }
-    
-    /** 
+
+    /**
      * {@inheritDoc}
      */
     @Override
-    public boolean getLongAddress() { return false;}
+    public boolean getLongAddress() {
+        return dccIsLong;
+    }
 
-    /** 
+    /**
      * {@inheritDoc}
      */
     @Override
-    public int getAddressNumber() {return 0;}
+    public int getAddressNumber() {
+        return dccAddress;
+    }
 
-    /** 
+    /**
      * {@inheritDoc}
      */
     @Override
-    public String getAddress() { return "";}
+    public String getAddress() {
+        if (dccAddress > 0 || dccIsLong) {
+            return Integer.toString(dccAddress) + (dccIsLong ? "L" : "S");
+        }
+        if (nid != null)
+            return nid.toString();
+        return "null";
+    }
+
+    private OlcbInterface getInterface() {
+        return iface;
+    }
+
+    /**
+     * Translates a (string) CV number into an address to read on the
+     *
+     * @param cvName CV address as provided to the various interface functions.
+     * @return memory space address to perform the read/write to.
+     */
+    private long getCvAddress(String cvName) {
+        int cvNum = Integer.parseInt(cvName);
+        return cvNum - 1;
+    }
+
+    class ProgTrackListener extends MessageDecoder {
+        @Override
+        public void handleProducerIdentified(ProducerIdentifiedMessage msg, Connection sender) {
+            if (!msg.getEventID().equals(IS_PROGRAMMINGTRACK_EVENT)) {
+                return;
+            }
+            if (msg.getSourceNodeID() == null) {
+                log.error("Found programming track with null source node.");
+                return;
+            }
+            if (msg.getSourceNodeID().getContents()[0] == 0) {
+                log.error("Found programming track with invalid source node: " + msg.getSourceNodeID().toString());
```

Comment:
Static analysis failes on this line. Replace it with:
```
log.error("Found programming track with invalid source node: {}", msg.getSourceNodeID().toString());
```

Reply:
thanks, fixed.